onchain/src/art_peace.cairo

```diff
@@ -976,5 +984,107 @@ pub mod ArtPeace {
             }
         }
     }
+
+    /// Internals
+    fn finalize_color_votes(ref self: ContractState) {
+        // Should we use a storage variable here ?
+        let nb_max: u8 = 3;
+
+        let day = self.day_index.read();
+        let votable_colors_count = self.votable_colors_count.read(day);
+
+        let mut max_scores: Felt252Dict<u32> = Default::default();
+        let mut max_indexes: Felt252Dict<u8> = Default::default();
+        let mut max_index: u8 = 0;
+
+        let mut votable_index: u8 = 1; // 0 means no vote
+
+        while votable_index <= votable_colors_count {
+            max_index = 0;
+            let mut index = votable_index;
+            let mut vote = self.color_votes.read((index, day));
+            // update max scores if needed
+            while max_index < nb_max {
+                let mut m: felt252 = max_index.into();
+                if (max_scores.get(m) < vote) {
+                    let mut old_vote = max_scores.get(m);
+                    let mut old_index = max_indexes.get(m);
+                    max_scores.insert(m, vote);
+                    max_indexes.insert(m, index);
+                    vote = old_vote;
+                    index = old_index;
+
+                    max_index += 1;
+                    // update max dict
+                    while max_index < nb_max {
+                        m = max_index.into();
+                        old_vote = max_scores.get(m);
+                        old_index = max_indexes.get(m);
+                        max_scores.insert(m, vote);
+                        max_indexes.insert(m, index);
+                        vote = old_vote;
+                        index = old_index;
+                        max_index += 1;
+                    };
+                };
+                max_index += 1;
+            };
+            votable_index += 1;
+        };
```

Comment:
This seems good to me, my only concern is something like this :
Color 1 : 5 votes
Color 2 : 8 votes
Color 3 : 7 votes
Color 4 : 5 votes
Color 5 : 4 votes

Color 1 & 4 would have the same 3rd most votes, but only Color 4 would get added to the palette. It might be better to be willing to add more to the palette if multiple share vote counts.
I think the best way to do that without changing your code too much would be to remove the `max_indexes` map, and when updating the palette, check if the vote count of the color >= the minimum value in `max_scores`.

Reply:
Ok I have updated source code to handle this case.
I have also added a test case with these votes to ensure we have the right behaviour.